### DISCUSSION TITLE
Understanding shap summary plot

### DISCUSSION THREAD

**Discussion by VickyVouk** [shap/shap#1595] [Q&A]
<img width="777" alt="Screen Shot 2020-11-17 at 11 56 21 PM" src="https://user-images.githubusercontent.com/35956507/99460691-9a5eb880-2930-11eb-90c9-68b7f0dabb6c.png">
I am trying to explain my results, but it seems hard to understand. On the picture I present only 2 variables, so that I do not overwhelm you. These variables are numbers of registered events in a country and my target variable is violence. Therefore, I would expect to see that the higher the number of registered events in assassinations the higher the violence score. However this plot shows that the smaller the number of registered events in assassinations (blue data on the right) the higher the violence? It does not make sense to me. On the other hand, the higher the number of registered events in assassinations (red data on the left) the lower the violence? No sense to me. And why do I see this tail on the left whereas on the right I see the data more concentrated? Does it mean something in particular or it is simply the distribution? The same holds for the other variable provided. Thank you very very much for the help. I also cit the expert @slundberg

**Comment by jaxondk** [comment]
Can you include some code? Is your problem classification or regression? Are you using the shap values for class 0 or class 1?

And the beeswarm plots do indeed show the distribution. Every dot corresponds to a data instance, and it's showing the SHAP value for that data instance. It's likely that those few dots you see to the left are outliers. You may want to look at the feature values there. What could be happening is that except for those outliers you are seeing the trend you would expect, as it looks like the right side of "attempts to assassinate" is actually purple-redish. Without those few instances that have very high assassination attempts, those might appear more red. 

Basically, you might be doing something wrong (but we cannot tell without more information), or you may simply need to massage the data and recognize the nuances going on in the plots. Also, try looking at multiple plots - I would look at the dependence plot for attempts to assassinate and see if you learn anything interesting.

**Reply by VickyVouk** [reply]
Hi @jaxondk . Thanks a lot for the reply. It is a regression problem. My target is violence (varying from 1.00 to 5.00 ) and the higher the score the higher the violence. And my features are events of violence registered in the USA in a week. So, my dataset is consisted of weeks - times series and I have the number of events registered per week for each feature(e.g. for the first week of 2020 I have 5 registered events of "attempt to assassinate". etc). I followed exactly the code here https://github.com/slundberg/shap so I do not believe there should be any code problem. The explanation you gave me with the outliers etc, I believe is the appropriate. I should look deeper on this data points, it should be interesting. Apart from the events features, I have also emotion features, and here I see a bigger problem, maybe you can help. 
<img width="683" alt="Screen Shot 2020-11-20 at 8 25 54 PM" src="https://user-images.githubusercontent.com/35956507/99841568-ad5ecc00-2b6e-11eb-9bcc-c2c7e1989eab.png">
In general, my emotion features'values vary from negative to positive numbers and they show the emotion that cause the events on a country. Looking on the picture, the variable Threaten-Emotion has a min value  -202.52 and a max 301.13 and a median 9.83 and we are talking for the country of US. I do not understand why blue points ( so very negative emotions of threaten) have a negative impact on violence and thus decrease violence. Similarly, I do not understand why red points (so positive emotions of threaten) have a positive impact on violence and thus increase violence. Shouldn't it be the other way round? The problem is that I see this pattern also with the other variables. I also normalised my values, so that I have not positive and negative, but it does not seem to be the issue. I really need help on this :)

**Reply by barnardp** [reply]
Not sure how relevant this is to the question here, but why does the color range used by summery plot only consider the values of the features for which the shapley values are being plotted? Would it make more sense to base the color range on the background dataset as this would allow one to better compare how the feature values effect the model relative to our background expectation. I would imagine this becomes particularly important if we want to compare summary plots across different parts of the test set, i.e. for time series data where we want to see if the distributions changes from one day to another etc.  

Expanding this thought a little more, if we've previously standardised our data according to the mean and deviation of our entire training set for training the model, but are then limited to using a background set which is only a small sample of our training set (and therefore has its own  unique mean and std deviation statistics), which of these sets of mean/deviation statistics should we use to color the shapley values? 

To try and understand this a little more, I've previously played around with the various choices by plotting PDPs of a linear model and seeing which sets, when used to calculate the model expected output and feature expected value, leads to the unique property that the PDP line passes though their intersection - From this I found that the property only holds true when the background set of the explainer is used by the PDP plot, suggesting that the mean/std deviation we initially used to normalise our data for training becomes somewhat irrelevant. Of course, if we were able to use our entire training set then this wouldn't even be an issue anymore but in cases where our training set is too large and we really are limited to using only a sample of the training set as our background, do you think it makes sense to using the mean/std deviation from our background sample to actually standardise the entire training set before training our model?? @slundberg @HughChen 

Any feedback would be much appreciated,
Pieter
